docs/examples/Fine-Tuning-Function-Calling.md

```diff
@@ -0,0 +1,273 @@
+
+Here's a brief comparison of the two:
+
+- **Full-weight fine-tuning**: it fine-tunes the entire model weights to better follow the instruction and examples in the training dataset. It is able to leverage the entire  model capacity, but it is expensive and prone to overfitting.
+- **LoRA fine-tuning**: it adapts the model to the task by adding a small number of parameters (0.1% to 1%) to the model, and train only those parameters. The new parameters are sufficient to capture the task-specific patterns and enable the model to generate the desired output. After training, we merge the new parameters into the original model to be used for inference. It is much more efficient than full-weight fine-tuning, and it is less prone to overfitting.
+
+Levanter provides good support for both methods. Therefore, we can easily try both methods and compare their results.
+
+#### Full-weight Fine-tuning
+
+We start with full-weight fine-tuning. Below is our configuration. Noteably:
+
+- The base model is `meta-llama/Llama-2-7b-hf`. It is set as the default value, so we don't need to specify it explicitly.
+- Batch size: We set the batch size to 128, which is the maximum batch size that can fit into a single TPUv3-8.
+- Learning rate: We compared the results with 3 epochs vs 2 epochs, and found that 2 epochs is sufficient to achieve the best results, while 3 epochs leads to overfitting.
+
+```yaml
+data: "gs://levanter-data/fine-tuning/gem-viggo/GEM_viggo_train.jsonl"
+data_cache_dir: "gs://levanter-data/tokenized/GEM_viggo_llama/"
+trainer:
+  wandb:
+    project: "levanter"
+    tags: ["viggo", "llama", "full-weight"]
+  mp: p=f32,c=bfloat16
+  train_batch_size: 128
+  num_train_steps: 80  # 5103 examples / 128 batch size * 2 epochs
+  tensor_parallel_axes: ["mlp", "heads"]
+optimizer:
+  learning_rate: 2E-5
+```
+
+The detailed instruction to run the training job can be found in the [Fine-Tuning documentation](../Fine-Tuning.md).
+Here is the command to run the training job on TPU:
+
+```bash
+gcloud compute tpus tpu-vm ssh finetune-32 --zone us-east1-d --worker=all \
+--command="WANDB_API_KEY=${YOUR WANDB TOKEN HERE} \
+HUGGING_FACE_HUB_TOKEN=${YOUR HF TOKEN HERE} \
+bash levanter/infra/run.sh python \
+levanter/examples/alpaca/alpaca.py \
+--config_path gs://<config-yaml-file> \
+--hf_save_path gs://<somewhere>"
+```
+
+Given the small dataset and high efficiency of Levanter, the entire training job completed quickly in only 21 min on a single TPUv3-8.
+
+Below shows an example where we compare the output from the original chat model and the fine-tuned model. We can see that the fine-tuned model is able to generate the correct function call and precisely capture the right attributes, while the original model is not able to do so.
```

Comment:
maybe give some numbers (baseline vs fine tuned) right before the example, just to set some expectations. Typical flow for something like this is `task overview -> dataset overview -> approach -> baseline results -> (exp -> results -> error analysis/example)*`